File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.java
class Solution {
    private boolean isPossible(int weights[],int maxW,int days) {
        int d = 1, currW = 0;
        for(int i = 0; i < weights.length; i++) {
            if(currW + weights[i] <= maxW) {
                currW += weights[i];
            } else {
                d++;
                currW = weights[i];
            }
        }
        if(d <= days) {
             return true;
        } else {
             return false;
        }
    }
    public int shipWithinDays(int[] weights, int days) {
        int high = 0;
        int low = Integer.MIN_VALUE;
        int n = weights.length;
        for(int i = 0; i < n; i++) {
            high += weights[i];
            if(weights[i] > low ) {
                low = weights[i];
            }
        }
        int ans = high;
        while(low <= high) {
            int mid = (low + high) / 2;
            if(isPossible(weights,mid,days)) {
                ans = mid;
                high = mid - 1;
            }else {
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: 1011-capacity-to-ship-packages-within-d-days/1011-capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    bool isPossible(vector<int>& weights,int maxW,int days) {
        int d = 1, currW = 0;
        for(auto w : weights) {
            if(w + currW <= maxW) {
                currW += w;
            }else {
                d++;
                currW = w;
            }
        }
        if(d <= days) {
            return true;
        }
        return false;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int low = INT_MIN,high = 0;
        for(auto w : weights) {
            high += w;
            low = max(low,w);
        }
        int ans;
        while(low <= high ) {
            int mid = low +( high - low )/ 2;
            if(isPossible(weights,mid,days)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }
};